server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -128,6 +128,23 @@ public RecoveryTarget(
         indexShard.recoveryStats().incCurrentAsTarget();
     }
 
+    private void refreshMultiFileWriter() {
+        // Sometimes we need to clear the downloaded data and start from scratch
+        // i.e. when we're recovering from a snapshot that's physically different to the
+        // source node index files. In that case we create a new MultiFileWriter using a
+        // different tempFilePrefix and close the previous writer that would take care of
+        // cleaning itself once all the outstanding writes finish.
+        multiFileWriter.close();
+        final String tempFilePrefix = RECOVERY_PREFIX + UUIDs.randomBase64UUID() + ".";
+        this.multiFileWriter = new MultiFileWriter(
+            indexShard.store(),
+            indexShard.recoveryState().getIndex(),
+            tempFilePrefix,
+            logger,
+            this::ensureRefCount
+        );
```

Comment:
Can we share this with the code in the constructor, looks identical - and is important that it is.